agent/agent_test.go

```diff
@@ -967,6 +970,135 @@ func TestAgent_AddServiceWithH2CPINGCheck(t *testing.T) {
+	keyFile := "../test/key/ourdomain_server.key"
+	caFile := "../test/ca/root.cer"
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	require.Equal(t, err, nil)
+
+	rootCertPool := x509.NewCertPool()
+	if caFile != "" {
+		caCert, err := os.ReadFile(caFile)
+		require.Equal(t, err, nil)
+		rootCertPool.AppendCertsFromPEM(caCert)
+	}
+
+	// Configure TLS to require and verify the client's certificate
+	config := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		ClientCAs:    rootCertPool,
+	}
+
+	l, err := tls.Listen("tcp", "127.0.0.1:0", config)
+	require.Equal(t, err, nil)
+	log.Printf("test tcp server listening on %s", l.Addr())
+	server := &testTCPServer{l: l}
+	go server.accept(t)
+
+	return server
+}
+
+// Close stops the server
+func (s *testTCPServer) close() {
+	atomic.StoreInt32(&s.stopped, 1)
+	if s.l != nil {
+		s.l.Close()
+	}
+}
+
+// Addr returns the address that this server is listening on.
+func (s *testTCPServer) Addr() net.Addr {
+	return s.l.Addr()
+}
+
+func (s *testTCPServer) accept(t *testing.T) error {
+	for {
+		conn, err := s.l.Accept()
+		if err != nil {
+			if atomic.LoadInt32(&s.stopped) == 1 {
+				t.Logf("test tcp echo server %s stopped", s.l.Addr())
+				return nil
+			}
+			t.Logf("test tcp echo server %s failed: %s", s.l.Addr(), err)
+			return err
+		}
+
+		t.Logf("test tcp echo server accepted connection from: %s\n", conn.RemoteAddr())
+		atomic.AddInt32(&s.accepted, 1)
+		atomic.AddInt32(&s.active, 1)
+
+		go func(c net.Conn) {
+			io.Copy(c, c)
+			atomic.AddInt32(&s.closed, 1)
+			atomic.AddInt32(&s.active, -1)
+		}(conn)
+	}
+}
+
+func TestAgent_AddServiceWithTCPTLSCheck(t *testing.T) {
+	t.Parallel()
+	dataDir := testutil.TempDir(t, "agent") // we manage the data dir
+	a := NewTestAgent(t, `
+		data_dir = "`+dataDir+`"
+		enable_agent_tls_for_checks = true
+		datacenter = "dc1"
+		tls {
+			defaults {
+				ca_file   = "../test/ca/root.cer"
+				cert_file = "../test/key/ourdomain_server.cer"
+				key_file  = "../test/key/ourdomain_server.key"
+			}
+		}
+	`)
+	defer a.Shutdown()
+	testrpc.WaitForTestAgent(t, a.RPC, "dc1")
+
+	// Create the temporary TCP socket server that the healthcheck will connect
+	// to.
+	testApp := newTestTCPServer(t)
+	defer testApp.close()
+
+	check := &structs.HealthCheck{
+		Node:    "foo",
+		CheckID: "arbitraryTCPServerTLSCheck",
+		Name:    "arbitraryTCPServerTLSCheck",
+		Status:  api.HealthCritical,
+	}
+
+	chkType := &structs.CheckType{
+		TCP:           testApp.Addr().String(),
+		TCPUseTLS:     true,
+		TLSServerName: "server.dc1.consul",
+		Interval:      5 * time.Second,
+	}
+
+	err := a.AddCheck(check, chkType, false, "", ConfigSourceLocal)
+	require.Equal(t, err, nil)
+
+	// Retry until the healthcheck is passing.
+	retry.Run(t, func(r *retry.R) {
+		status := getCheck(a, "arbitraryTCPServerTLSCheck")
+		if status.Status != api.HealthPassing {
+			r.Fatalf("bad: %v", status.Status)
+		}
+	})
+}
+
```

Comment:
The implementation you have here is great for debugging your implementation. However, for the final test case, we should go with something smaller that uses more of the standard library:

```suggestion
func startMockTLSServer(t *testing.T) (addr string, closeFunc func()) {
	// Load certificates
	cert, err := tls.LoadX509KeyPair("../test/key/ourdomain_server.cer", "../test/key/ourdomain_server.key")
	require.NoError(t, err)

	// Create a certificate pool
	rootCertPool := x509.NewCertPool()
	caCert, err := os.ReadFile("../test/ca/root.cer")
	require.NoError(t, err)
	rootCertPool.AppendCertsFromPEM(caCert)

	// Configure TLS
	config := &tls.Config{
		Certificates: []tls.Certificate{cert},
		ClientAuth:   tls.RequireAndVerifyClientCert,
		ClientCAs:    rootCertPool,
	}

	// Start TLS server
	ln, err := tls.Listen("tcp", "127.0.0.1:0", config)
	require.NoError(t, err)

	go func() {
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			go func(c net.Conn) {
				io.Copy(c, c)
				c.Close()
			}(conn)
		}
	}()

	return ln.Addr().String(), func() { ln.Close() }
}

func TestAgent_AddServiceWithTCPTLSCheck(t *testing.T) {
	t.Parallel()
	dataDir := testutil.TempDir(t, "agent")
	a := NewTestAgent(t, `
		data_dir = "`+dataDir+`"
		enable_agent_tls_for_checks = true
		datacenter = "dc1"
		tls {
			defaults {
				ca_file   = "../test/ca/root.cer"
				cert_file = "../test/key/ourdomain_server.cer"
				key_file  = "../test/key/ourdomain_server.key"
			}
		}
	`)
	defer a.Shutdown()
	testrpc.WaitForTestAgent(t, a.RPC, "dc1")

	// Start mock TCP+TLS server
	addr, closeServer := startMockTLSServer(t)
	defer closeServer()

	check := &structs.HealthCheck{
		Node:    "foo",
		CheckID: "arbitraryTCPServerTLSCheck",
		Name:    "arbitraryTCPServerTLSCheck",
		Status:  api.HealthCritical,
	}

	chkType := &structs.CheckType{
		TCP:           addr,
		TCPUseTLS:     true,
		TLSServerName: "server.dc1.consul",
		Interval:      5 * time.Second,
	}

	err := a.AddCheck(check, chkType, false, "", ConfigSourceLocal)
	require.NoError(t, err)

	// Retry until the healthcheck is passing.
	retry.Run(t, func(r *retry.R) {
		status := getCheck(a, "arbitraryTCPServerTLSCheck")
		if status.Status != api.HealthPassing {
			r.Fatalf("bad: %v", status.Status)
		}
	})
}

```